Add multilayer VIA support to Hasu usb_usb

jshuf opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

Adds VIA support to the Hasu usb_usb converter.  'multilayer' comes from some additional complexity above and beyond most VIA ports - in order to fit more than one dynamic keymap layer into the atmega32u4 EEPROM, the converter's default scheme of using USB HID codes to directly index a 16x16 scan matrix is replaced with an intermediate mapping.  This mapping translates from the smaller (142-code) set of USB HID codes actually used by the existing usb_usb layout macros, to cells in a smaller 9x16 scan matrix.  This reduced matrix size allows three dynamic keymap layers to be stored in the atmega32u4 EEPROM.  A more detailed explanation can be found in the comments in keymaps/via/keymap.c.

All code changes to support this are ifdef'd under VIA_ENABLE, and hence should have no impact on existing usage of the QMK usb_usb converter code.  Because the LAYOUT macro for this VIA implementation has scan matrix dimensions that differ from the others, it lives in keymaps/via/keymap.c to avoid confusion with existing LAYOUT macros and info.json.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* None

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -39,6 +39,66 @@ extern "C" {
 }
 
+/* see keymaps/via/keymap.c for VIA implementation details */
+#ifdef VIA_ENABLE
```
why is there via specific implementation here?

drashna replied:
That doesn't really answer the question, though.  Why does there need to be via specific handling of this? 

VIA should be unaware of the hardware, here.   The dynamic keymap feature, which via leverages changes the keymap, and that reads from the matrix config.

Eg, there is not really a good reason to have via specific code if the matrix is properly implemented, as far as I can see.

jshuf replied:
I'm still not 100% sure I understand the issue, but is it that you'd prefer a different ifdef to wrap that code?

There are two scan matrices (the original, which supports translation of all 256 USB keycodes but only one layer, and the new one I've added, which supports fewer keycodes in order to allow for three layers).  Both matrices are implemented correctly, and the USB-to-USB converter can support both - but you have to choose one.  Since only one of the two matrices supports multiple layers, it made sense to me to put it under an ifdef so that you can make that compile-time choice.  If the issue you're getting at is that using VIA_ENABLE is too specific a toggle in this hardware-specific code to make that choice (since any dynamic keymap user might want to enable it) then I could put the smaller one under some other ifdef and then enable that new ifdef only in the via keymap, which would take the VIA ifdef out of custom_matrix.cpp.

fauxpark commented:
I'd like to convert this board to Custom Matrix Lite first - this change should still be doable after that, but hopefully with a little less boilerplate.

FREEWING-JP commented:
This PR is very helpful to me for Reduce program size .
http://www.neko.ne.jp/~freewing/hardware/qmk_usb_converter_max3421e_host_shield/

jshuf commented:
Still active (at least as far as I'm concerned).

jshuf commented:
Still active.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______, EEP_RST,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,           ______,  ______,                                    ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,           ______,  ______,               ______,              ______,  ______,  ______,  ______,     ______,  ______,
+     DEBUG,  ______,  ______,  ______,  ______,          ______,         ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,           ______,  ______,     ______,  ______
```
```suggestion
    DB_TOGG,  ______,  ______,  ______,  ______,          ______,         ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,           ______,  ______,     ______,  ______
```

jshuf replied:
Replaced DEBUG with QK_DEBUG_TOGGLE in the latest push.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+                      ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,
+     RESET,           ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,               ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,
+    ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______, EEP_RST,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
```
```suggestion
    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  EE_CLR,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
```

jshuf replied:
Replaced with QK_CLEAR_EEPROM in latest push.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+    ),
+    [2] = LAYOUT_via(
+                      ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,
+     RESET,           ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,               ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,
```
```suggestion
   QK_BOOT,           ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,               ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,
```

jshuf replied:
Replaced with QK_BOOT in latest push.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,   KC_Y,   KC_U,   KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC,          KC_BSLS,     KC_DEL,  KC_END,  KC_PGDN,    KC_P7,   KC_P8,   KC_P9,   KC_PPLS,    KC_MENU, KC_UNDO,
+    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,   KC_H,   KC_J,   KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_NUHS, KC_ENT,                                    KC_P4,   KC_P5,   KC_P6,   KC_PCMM,    KC_SLCT, KC_COPY,
+    KC_LSFT, KC_NUBS, KC_Z,    KC_X,    KC_C,    KC_V,   KC_B,   KC_N,   KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RO,   KC_RSFT,              KC_UP,               KC_P1,   KC_P2,   KC_P3,   KC_PEQL,    KC_EXEC, KC_PSTE,
+    KC_LCTL, KC_LGUI, KC_LALT, KC_MHEN, KC_HANJ,         KC_SPC,         KC_HAEN, KC_HENK, KC_KANA, KC_RALT, KC_RGUI, MO(1),   KC_RCTL,     KC_LEFT, KC_DOWN, KC_RGHT,    KC_P0,            KC_PDOT, KC_PENT,    KC_FIND, KC_CUT
```
```suggestion
    KC_LCTL, KC_LGUI, KC_LALT, KC_INT5, KC_LNG2,         KC_SPC,         KC_LNG1, KC_INT4, KC_INT2, KC_RALT, KC_RGUI, MO(1),   KC_RCTL,     KC_LEFT, KC_DOWN, KC_RGHT,    KC_P0,            KC_PDOT, KC_PENT,    KC_FIND, KC_CUT
```

jshuf replied:
Replaced as suggested in the latest push.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+    [0] = LAYOUT_via(
+                      KC_F13,  KC_F14,  KC_F15,  KC_F16, KC_F17, KC_F18, KC_F19,  KC_F20,  KC_F21,  KC_F22,  KC_F23,  KC_F24,
+    KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,  KC_F5,  KC_F6,  KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,               KC_PSCR, KC_SLCK, KC_PAUS,    KC_VOLD, KC_VOLU, KC_MUTE, KC_PWR,     KC_HELP,
+    KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,   KC_6,   KC_7,   KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_JYEN, KC_BSPC,     KC_INS,  KC_HOME, KC_PGUP,    KC_NLCK, KC_PSLS, KC_PAST, KC_PMNS,    KC_STOP, KC_AGIN,
```
```suggestion
    KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,   KC_6,   KC_7,   KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_INT3, KC_BSPC,     KC_INS,  KC_HOME, KC_PGUP,    KC_NUM,  KC_PSLS, KC_PAST, KC_PMNS,    KC_STOP, KC_AGIN,
```

jshuf replied:
Replaced as suggested in the latest push.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,127 @@
+     */
+    [0] = LAYOUT_via(
+                      KC_F13,  KC_F14,  KC_F15,  KC_F16, KC_F17, KC_F18, KC_F19,  KC_F20,  KC_F21,  KC_F22,  KC_F23,  KC_F24,
+    KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,  KC_F5,  KC_F6,  KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,               KC_PSCR, KC_SLCK, KC_PAUS,    KC_VOLD, KC_VOLU, KC_MUTE, KC_PWR,     KC_HELP,
```
```suggestion
    KC_ESC,           KC_F1,   KC_F2,   KC_F3,   KC_F4,  KC_F5,  KC_F6,  KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,               KC_PSCR, KC_SCRL, KC_PAUS,    KC_VOLD, KC_VOLU, KC_MUTE, KC_PWR,     KC_HELP,
```

jshuf replied:
Replaced as suggested in latest push.

jshuf commented:
Updated PR to reflect changes to master, and replaced VIA_ENABLE with COMPRESS_MATRIX, to better reflect that the changes to the keymap are not VIA-specific.

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,129 @@
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______, QK_CLEAR_EEPROM,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,           ______,  ______,                                    ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,           ______,  ______,               ______,              ______,  ______,  ______,  ______,     ______,  ______,
+     QK_DEBUG_TOGGLE,  ______,  ______,  ______,  ______,          ______,         ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,           ______,  ______,     ______,  ______
```
```suggestion
     DB_TOGG,  ______,  ______,  ______,  ______,          ______,         ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,           ______,  ______,     ______,  ______
```

fauxpark commented on keyboards/converter/usb_usb/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,129 @@
+                      ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,
+     QK_BOOT,           ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,               ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,
+    ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______,  ______,  ______,  ______,  ______,  ______,  ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
+    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______, QK_CLEAR_EEPROM,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
```
```suggestion
    ______,  ______,  ______,  ______,  ______,  ______, ______, ______, ______,  ______, EE_CLR,  ______,  ______,           ______,      ______,  ______,  ______,     ______,  ______,  ______,  ______,     ______,  ______,
```

fauxpark commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -228,11 +306,13 @@ extern "C" {
     }
 
-    void led_set(uint8_t usb_led) {
+    void led_set(uint8_t usb_led)
+    {
```
```suggestion
    void led_set(uint8_t usb_led) {
```

fauxpark commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -228,11 +306,13 @@ extern "C" {
         led_set_kb(usb_led);
     }
-}
+
+};
```
```suggestion
}
```

fauxpark commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -58,31 +118,36 @@ extern "C" {
+USBHub hub2(&usb_host);
 
-extern "C" {
+
+extern "C"
+{
```
```suggestion
extern "C" {
```

fauxpark commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -58,31 +118,36 @@ extern "C" {
-HIDBoot<HID_PROTOCOL_KEYBOARD> kbd2(&usb_host);
-HIDBoot<HID_PROTOCOL_KEYBOARD> kbd3(&usb_host);
-HIDBoot<HID_PROTOCOL_KEYBOARD> kbd4(&usb_host);
+HIDBoot<HID_PROTOCOL_KEYBOARD>    kbd1(&usb_host);
+HIDBoot<HID_PROTOCOL_KEYBOARD>    kbd2(&usb_host);
+HIDBoot<HID_PROTOCOL_KEYBOARD>    kbd3(&usb_host);
+HIDBoot<HID_PROTOCOL_KEYBOARD>    kbd4(&usb_host);
```
```suggestion
HIDBoot<HID_PROTOCOL_KEYBOARD> kbd1(&usb_host);
HIDBoot<HID_PROTOCOL_KEYBOARD> kbd2(&usb_host);
HIDBoot<HID_PROTOCOL_KEYBOARD> kbd3(&usb_host);
HIDBoot<HID_PROTOCOL_KEYBOARD> kbd4(&usb_host);
```

fauxpark commented on keyboards/converter/usb_usb/custom_matrix.cpp:
```diff
@@ -58,31 +118,36 @@ extern "C" {
-#define ROW(code)      (((code) & ROW_MASK) >> 4)
-#define COL(code)      ((code) & COL_MASK)
-#define ROW_BITS(code) (1 << COL(code))
+#define CODE(row, col)  (((row) << 4) | (col))
+#define ROW(code)       (((code) & ROW_MASK) >> 4)
+#define COL(code)       ((code) & COL_MASK)
+#define ROW_BITS(code)  (1 << COL(code))
```
```suggestion
#define CODE(row, col) (((row) << 4) | (col))
#define ROW(code)      (((code) & ROW_MASK) >> 4)
#define COL(code)      ((code) & COL_MASK)
#define ROW_BITS(code) (1 << COL(code))
```

jshuf commented:
The remaining CI failures are outside the scope of this PR - they're in userspace code, and are failing to compile in master.

bbecker-inlogik commented:
@jshuf In an attempt to free up as much EEPROM as possible, In my local code base, I added a keymaps/via_ansi folder which removed 34 keys in total ( F13-F24, International keys, legacy key codes ).  So that makes 108 HID keycodes for a US ANSI layout, which could fit into 16x7 with 4 spare.  
How would I go about removing 2 rows from hidindex?  I suppose I don't fully understand the meaning of val = ((row) << 4) + (col)) and how that fits in with the various codes ie. 0x5D.  Thanks for your PR. I'd like to see this get included in the code base.

bbecker-inlogik commented:
Merge conflicts fixed in attached custom_matrix.cpp file.
[custom_matrix.zip](https://github.com/qmk/qmk_firmware/files/12678172/custom_matrix.zip)

jshuf commented:
@bbecker-inlogik thanks for the code changes - I'll try to integrate and resubmit when I have time.  There have been other QMK changes since the last review that I need to factor back into this code.  In the meantime...

If foo is the HID code emitted by the keyboard and received by the converter, then hidindex[matrixindex[foo]] is the HID code that the converter will send to the host.  That extra level of indirection is what allows us to reduce the size of each VIA layer in EEPROM.
So, to make 16x7 work in your local code base, you'd want to do at least these things (there may be more, I haven't looked at this in a while, but this should get you started):
- in via_ansi/config.h, set MATRIX_ROWS to 7
- make hidindex an array of length 112, and make sure it includes all 108 of the HID codes you want, filling the remaining entries of the array with 0x00.  The order in which you enumerate the 108 HID codes doesn't really matter here.
- matrixindex remains an array of length 256, representing all 256 HID codes that could theoretically be received by the converter.  Each value of this matrix is an index into the hidindex array.  So, for each of those 256 HID codes, matrixindex will contain either XBAD (meaning that the HID code won't be supported) or it will contain the index 0x{R}{C}, where (R<<4) + C is the index into hidindex where you've stored the HID code you want the converter to send to the host.
- in via_ansi/keymap.c, the LAYOUT_via macro and keymaps[][][] would need to be modified to correspond to your 16x7 array - this is what would initialize the layers in VIA.

sigprof commented:
The matrix compression feature could be useful even without VIA (the current implementation uses 400 bytes for the data tables, but reduces each layer by 224 bytes, so it may be a win even starting from 2 layers, depending on the code size increase).  So the implementation probably should not be coupled with `VIA_ENABLE`.

One major issue is that adding an option which changes the matrix layout makes all layouts which are currently defined in `info.json` invalid — all matrix locations in there must be adjusted to match the new virtual matrix.  The solutions that I could think of are:
1. Use the compressed matrix unconditionally; change all layout definitions in `info.json` accordingly.  This makes the converter work mostly like “normal” keyboards (no weird extra options), but might break things for users of strange keyboards which could send keycodes outside of the normal set (however, I'm not sure whether such USB keyboards actually exist in the wild, so this might not be a problem in practice).
2. Split the board into multiple revisions — one with the existing uncompressed matrix and layout definitions (and no VIA support), another with the compressed matrix and updated layout definitions.  Unfortunately, this board already has 3 subrevisions, which would then need to be duplicated for the uncompressed/compressed revisions.

sigprof commented:
In any case, changes like this are rather major, and need to go to the `develop` branch, not directly to `master`.

sigprof commented:
Also you should think about converting the matrix to the “lite” custom matrix — even though it would have some overhead due to the debounce logic (the debounce itself could be disabled, but the `matrix` and `raw_matrix` arrays would still remain separate), the compressed matrix would take just 18 bytes of RAM, which would probably be an acceptable price for using more generic code (and might actually end up being faster than the custom `matrix_get_row()` implementation).  And changing the code to convert the HID reports into bit arrays all at once would allow dropping the `hidindex` table (only `matrixindex` would be needed to determine the bit position for each pressed key).

tzarc commented:
As of August 26, 2024, `qmk/qmk_firmware` is no longer accepting VIA-enabled keymaps as these have now transitioned to a repository under the VIA team's control.

As you've submitted a PR containing `via` or VIA-enabled keymap(s), this is your notice that they should be removed from this PR. You should now submit a secondary PR to the [VIA QMK Userspace](https://github.com/the-via/qmk_userspace_via) repository with your associated `via` or VIA-enabled keymaps instead.
